src/ui/public/visualize/visualization.js

```diff
@@ -115,50 +116,64 @@ uiModules
 
-        const renderFunction = _.debounce(() => {
-          const container = getVisContainer();
-          if (!container) return;
-          $scope.vis.size = [container.width(), container.height()];
-          const status = getUpdateStatus($scope);
-          visualization.render($scope.visData, status)
-            .then(() => {
-            // renderComplete
-              $scope.$emit('renderComplete');
-              $el.trigger('renderComplete');
+        const dispatchCustomEvent = (name) => {
+          // we're using the native events so that we aren't tied to the jQuery custom events,
+          // otherwise we have to use jQuery(element).on(...) because jQuery's events sit on top
+          // of the native events per https://github.com/jquery/jquery/issues/2476
+          $el[0].dispatchEvent(new CustomEvent(name, { bubbles: true }));
+        };
+
+        const render$ = Observable.create(observer => {
+          $scope.$on('render', () => {
+            observer.next({
+              vis: $scope.vis,
+              visData: $scope.visData,
+              container: getVisContainer(),
             });
-          $scope.$apply();
-        }, 100);
-
-        $scope.$on('render', () => {
-          if (!$scope.vis || !$scope.vis.initialized || ($scope.vis.type.requiresSearch && !$scope.visData)) {
-            return;
-          }
-          $scope.addLegend = $scope.vis.params.addLegend;
-          $scope.vis.refreshLegend++;
-          $timeout(renderFunction);
+          });
         });
 
+        const success$ = render$
+          .do(() => {
+            dispatchCustomEvent('renderStart');
+          })
+          .filter(({ vis, visData, container }) => vis && vis.initialized && container && (!vis.type.requiresSearch || visData))
+          .do(({ vis }) => {
+            $scope.addLegend = vis.params.addLegend;
+            vis.refreshLegend++;
+          })
+          .debounceTime(100)
+          .switchMap(async ({ vis, visData, container }) => {
+            vis.size = [container.width(), container.height()];
+            const status = getUpdateStatus($scope);
+            const renderPromise = visualization.render(visData, status);
+            $scope.$apply();
+            return renderPromise;
+          });
+
+        const requestError$ = render$.filter(({ vis }) => vis.requestError);
+
+        const renderSubscription = Observable.merge(success$, requestError$)
+          .subscribe(() => {
+            window.requestAnimationFrame(() => {
+              $scope.$emit('renderComplete');
+              dispatchCustomEvent('renderComplete');
+            });
+          });
+
         $scope.$on('$destroy', () => {
           resizeChecker.destroy();
           visualization.destroy();
+          renderSubscription.unsubscribe();
         });
 
         if (!$scope.vis.visualizeScope) {
           $scope.$watchGroup(['visData', 'vis.params'], () => {
             $scope.$emit('render');
```

Comment:
do we still need these events? If not, it seems like an unnecessary indirection and the `render$` could listen to these param-changes directly.

Reply:
We still need the `render` event in general in case the directive will be notified from the outside that it needs to rerender, e.g. when visualize got new data. Thus I think we can safely use the event here instead of calling the observer directly. That way we have at least the same way if triggering a render in all places.